Disabling users in WebUI

Faldon opened the pull request:
# Ability to disable users in Web UI
## Base for implementing #283 
- Implemented an action menu for each user like in the files section
- Added the menu items delete and disable/enable (based on the current state of the user)
- Added a Disabled Group in the sidebar and a disbaled users count
## TODO:
- ~~When enabling a user, site needs a reload because the groups won't be filled with users anymore.
  The other way around seems to work fine.~~ (Edit: I just fired ab my local test server after the weekend and this strange behaviour was gone. Maybe a cache problem?)

Commit:
Implemented visual feedback if a user is disabled in admin user menu.

Commit:
Implemented visuals for enabling/disabling user from admin user list.
Added the controller functions for enabling/disabling a user.

Commit:
Added the route for changing user status (enabled/disabled) and added an additional route handler in the user controller.
 Finished the visuals to reflect current user status and changed user status respectively.

Commit:
Changed the single icon for enabling/disabling a user into a menu where deletion and state toggling of a user is selectable.

Commit:
Added displaying of disabled user count.
Improved style of user action menu.

Faldon committed:
Added proper counting of disabled users.
Removed visual indicator for disabled users.

MorrisJobke commented:
* [ ] PNGs are not needed anymore on master
* [ ] optimize the SVGs (use either `svgo` or `scour` to do so)

MorrisJobke commented on settings/Controller/UsersController.php:
```diff
@@ -489,6 +490,135 @@ public function destroy($id) {
+		} else {
+			return new DataResponse(
+				array(
+					'status' => 'error',
```
Please use HTTP status codes for the response instead of a status flag.

MorrisJobke replied:
Ah. You just used the same style as in the code above. Okay. Then we could keep this, but should clean this up once we touch all the code here again.

MorrisJobke commented:
We already have the menu CSS in place. This is for now only implemented in the sidebar, but should be reusable: https://docs.nextcloud.org/server/9/developer_manual/app/css.html#menus

![bildschirmfoto 2016-07-18 um 09 31 21](https://cloud.githubusercontent.com/assets/245432/16908302/a8634e38-4cca-11e6-933f-3a12a571e9b5.png)


Could you check this out? And also we usually use px instead of em. Maybe this already helps to make this work.

MorrisJobke commented:
But nevertheless: Thanks for your contribution :)

Faldon commented:
@MorrisJobke I changed my additions to the css from em to px. There are already some declarations in em, but I would rather look into it as separate task (redesigning settings maybe, than we also can change the code for the http status codes as you mentioned above.)

The last mentioned problem (not displaying user) was a bug that seems to be present already. In the template resides a single tr for the javascript, that should be kept on reloading the user list but wasn't.

Commit:
Moved pseudo-group detection for disabled users from frontend to the controller.
Changed units for newly introduced css values from em to px.
Removed unnecessary png and optimized svg with scour.
Changed the userlist template to display the user action menu with correct width.

MorrisJobke commented on settings/Controller/UsersController.php:
```diff
@@ -245,15 +240,18 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
 
 			foreach ($batch as $user) {
-				$users[] = $this->formatUserForIndex($user);
+				if( ($gid!=='disabledUsers' && $user->isEnabled()) ||
```
Missing spaces around `!==`

MorrisJobke commented on settings/Controller/UsersController.php:
```diff
@@ -266,7 +264,7 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
 
 			// Set the $gid parameter to an empty value if the subadmin has no rights to access a specific group
-			if($gid !== '' && !in_array($gid, $subAdminOfGroups)) {
+			if($gid !== '' && $gid!== 'disabledUsers' && !in_array($gid, $subAdminOfGroups)) {
```
missing space before `!==`

MorrisJobke commented on settings/Controller/UsersController.php:
```diff
@@ -291,7 +289,11 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
 					$subAdminOfGroups
 				));
-				$users[] = $this->formatUserForIndex($user, $userGroups);
+				if( ($gid!=='disabledUsers' && $user->isEnabled()) ||
```
missing space around `!==`

MorrisJobke commented on settings/js/users/groups.js:
```diff
@@ -216,7 +216,7 @@ GroupList = {
 		GroupList.activeGID = gid;
 		UserList.empty();
-		UserList.update(gid);
+		UserList.update(gid=='_everyone'?'':gid);
```
please use `===` and spaces around the operators (`===`, `?`, `:`) for better readability.

MorrisJobke commented:
Beside that it already looks quite good :)

MorrisJobke commented:
For better user experience we should make the loginException extend from HintException:

```php
use OC\HintException;

class LoginException extends HintException {
}
```

@Faldon Could you add this?

@LukasReschke I don't see a risk for leaking info here. Every time the login exception is thrown it also uses a translated message.

@Faldon Beside that it works really nice 👍

Faldon commented:
I'm still struggling with the code style. -.- Fixes are on their way.
@MorrisJobke Currently, the LoginController seems to not throw any Exception. Or am I looking at the wrong place?

Commit:
Style fixes for better readability and coding style conformity.

MorrisJobke commented:
> @MorrisJobke Currently, the LoginController seems to not throw any Exception. Or am I looking at the wrong place?

Correct. Then all is fine. This is the old way to implement it but it is consistent in that class. So leave it as it is :)

MorrisJobke commented:
@Faldon There are conflicts that doesn't allow merging. Could you try to resolve those conflicts?

Commit:
Merge branch 'master' into ui_enhancements-conflicts
Resolving conflicts for PR #413

Faldon commented:
Hmpf...
Merging the master branch back appearantly was not a good idea... ;-( Sorry for that. But hopefully, the merge conflicts are resolved now...

jancborchardt commented:
@Faldon I just noticed the added icons. Instead please use the classes `.icon-add` and `.icon-delete` with the respective text: »Add user« or »Delete user« (or »Disable«)

Faldon commented:
@jancborchardt Should I extend the icons style sheet for the disabling user icon then?

jancborchardt commented:
@Faldon we can use `.icon-close` for that. It’s similar to »Delete« but not a trash can. And it’s important there’s text next to it. :)

Commit:
Changed the icons for enabling, disabling and deleting a user in the action menu.

Faldon commented:
@jancborchardt Ok, I did the changes for the action menu. However, I left the code style for the "more" icon to comply with the other icons in the tr that are visible on hover.

MorrisJobke commented:
The conflicts doesn't seem to be trivial. I will try to sort this out. Sorry for having this laying around that long 😢

MorrisJobke commented:
Rebase on master is in #1234

@Faldon Is this okay for you? (that I take over this PR)

MorrisJobke commented:
> @Faldon Is this okay for you? (that I take over this PR)

I will close this here for now.

MorrisJobke commented:
Superseeded by #1234
